src/ruby/puppetserver-lib/puppet/server/master.rb

```diff
@@ -66,13 +72,19 @@ def handleRequest(request)
   end
 
   def compileCatalog(request_data)
-    @catalog_compiler.compile(convert_java_args_to_ruby(request_data))
+    Puppet.override(lookup_key_recorder: lookup_key_recorder) do
+      @catalog_compiler.compile(convert_java_args_to_ruby(request_data))
+    end
   end
 
   def compileAST(compile_options)
     Puppet::Server::ASTCompiler.compile(convert_java_args_to_ruby(compile_options))
   end
 
+  def lookup_key_recorder
```

Comment:
should put a verb in this name so that it's clear that it's making a new one each time instead of just some saved instance variable